lib/connections/limiter.go

```diff
@@ -12,53 +12,197 @@ import (
 	return l
 }
 
-func (lim *limiter) newReadLimiter(r io.Reader, isLAN bool) io.Reader {
-	return &limitedReader{reader: r, limiter: lim, isLAN: isLAN}
+// This function sets limiters according to corresponding DeviceConfiguration
+func (lim *limiter) setLimitsForDevice(v config.DeviceConfiguration, readLimiter, writeLimiter *rate.Limiter) {
+	// The rate variables are in KiB/s in the config (despite the camel casing
+	// of the name). We multiply by 1024 to get bytes/s.
+	if v.MaxRecvKbps <= 0 {
+		readLimiter.SetLimit(rate.Inf)
+	} else {
+		readLimiter.SetLimit(1024 * rate.Limit(v.MaxRecvKbps))
+	}
+
+	if v.MaxSendKbps <= 0 {
+		writeLimiter.SetLimit(rate.Inf)
+	} else {
+		writeLimiter.SetLimit(1024 * rate.Limit(v.MaxSendKbps))
+	}
+}
+
+// This function handles removing, adding and updating of device limiters.
+// Pass pointer to avoid copying. Pointer already points to copy of configuration
+// so we don't have to worry about modifying real config.
+func (lim *limiter) processDevicesConfiguration(from, to *config.Configuration) {
+
+	devicesToRemove := make(map[protocol.DeviceID]bool)
+	// Get all devices that were previously in map, these are candidates for removal
+	for _, v := range from.Devices {
+		if v.DeviceID != to.MyID {
+			devicesToRemove[v.DeviceID] = true
+		}
+	}
+
+	// Mark devices which should not be removed, create new limiters if needed and assign new limiter rate
+	for _, v := range to.Devices {
+		if v.DeviceID == to.MyID {
+			// This limiter was created for local device. Should skip this device
+			continue
+		}
+
+		readLimiter, okR := lim.getDeviceReadLimiter(v.DeviceID)
+		// Device has not been removed or added, this is no longer a candidate for removal
+		if okR {
+			devicesToRemove[v.DeviceID] = false
+		}
+
+		writeLimiter, okW := lim.getDeviceWriteLimiter(v.DeviceID)
+		if okW {
+			if devicesToRemove[v.DeviceID] {
+				// Device has not been removed or added, we should've
+				// already marked it as false in devicesToRemove
+				panic("broken symmetry in device read/write limiters")
+			}
+		}
+
+		// There was no limiter for this ID in map.
+		// This means that we added this device and we should create new limiter
+		if !okR && !okW {
+			readLimiter = rate.NewLimiter(rate.Inf, limiterBurstSize)
+			writeLimiter = rate.NewLimiter(rate.Inf, limiterBurstSize)
+
+			lim.setDeviceLimiters(v.DeviceID, writeLimiter, readLimiter)
+		} else if !okR || !okW {
+			// One of the read/write limiters is not present while the
+			// corresponding write/read one exists. Something has gone wrong.
+			panic("broken symmetry in device read/write limiters")
+		}
+
+		// limiters for this device are created so we can store previous rates for logging
+		previousReadLimit := readLimiter.Limit()
+		previousWriteLimit := writeLimiter.Limit()
+
+		l.Debugf("okR: %t, okW: %t, prevWriteLim: %d, prevReadLim: %d", okR, okW, previousReadLimit, previousWriteLimit)
+
+		lim.setLimitsForDevice(v, readLimiter, writeLimiter)
```

Comment:
Is it not better not to do this if the limits are the same as previously?
As in, move this after the if?